mass_mailing_custom_unsubscribe/models/mail_unsubscription.py

```diff
@@ -0,0 +1,120 @@
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+
+from odoo import _, api, fields, models
+from odoo.addons.mass_mailing.models.mass_mailing import \
+    MASS_MAILING_BUSINESS_MODELS
+from .. import exceptions
+
+
+class MailUnsubscription(models.Model):
+    _name = "mail.unsubscription"
+    _description = "Mail unsubscription"
+    _inherit = "mail.thread"
+    _rec_name = "date"
+    _order = "date DESC"
+
+    date = fields.Datetime(
+        default=lambda self: self._default_date(),
+        required=True)
+    email = fields.Char(
+        required=True)
+    action = fields.Selection(
+        selection=[
+            ("subscription", "Subscription"),
+            ("unsubscription", "Unsubscription"),
+            ("blacklisting", "Blacklisting"),
+            ("de_blacklisting", "De-blacklisting"),
+        ],
+        required=True,
+        default="unsubscription",
+        help="What did the (un)subscriber choose to do.",
+    )
+    mass_mailing_id = fields.Many2one(
+        "mail.mass_mailing",
+        "Mass mailing",
+        required=True,
+        help="Mass mailing from which he was unsubscribed.")
+    unsubscriber_id = fields.Reference(
+        lambda self: self._selection_unsubscriber_id(),
+        "(Un)subscriber",
+        help="Who was subscribed or unsubscribed.")
+    mailing_list_ids = fields.Many2many(
+        comodel_name="mail.mass_mailing.list",
+        string="Mailing lists",
+        help="(Un)subscribed mass mailing lists, if any.",
+    )
+    reason_id = fields.Many2one(
+        "mail.unsubscription.reason",
+        "Reason",
+        ondelete="restrict",
+        help="Why the unsubscription was made.")
+    details = fields.Text(
+        help="More details on why the unsubscription was made.")
+    details_required = fields.Boolean(
+        related="reason_id.details_required")
+    metadata = fields.Text(
+        readonly=True,
+        help="HTTP request metadata used when creating this record.",
+    )
+
+    def map_mailing_list_models(self, models):
+        model_mapped = []
+        for model in models:
+            if model == 'mail.mass_mailing.list':
+                model_mapped.append(('mail.mass_mailing.contact', model))
+            else:
+                model_mapped.append((model, model))
+        return model_mapped
+
+    @api.model
+    def _default_date(self):
+        return fields.Datetime.now()
+
+    @api.model
+    def _selection_unsubscriber_id(self):
+        """Models that can be linked to a ``mail.mass_mailing``."""
+        model = self.env['ir.model'].search(
+            [('model', 'in', MASS_MAILING_BUSINESS_MODELS)]).mapped('model')
+        return self.map_mailing_list_models(model)
+
+    @api.multi
+    @api.constrains("action", "reason_id")
+    def _check_reason_needed(self):
+        """Ensure reason is given for unsubscriptions."""
+        for one in self:
+            if one.action == "unsubscription" and not one.reason_id:
```

Comment:
Should we change it to this now?:

```suggestion
            if one.action in {"unsubscription", "blacklist_add"} and not one.reason_id:
```